// @vfx-js/core/lib/esm/backbuffer.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Backbuffer_width, _Backbuffer_height, _Backbuffer_pixelRatio, _Backbuffer_buffers;
import * as THREE from "three";
import { getGLRect } from "./gl-rect";
/**
 * A class to manage initialization and double-buffring of the backbuffer.
 * @internal
 */
export class Backbuffer {
    constructor(width, height, pixelRatio) {
        _Backbuffer_width.set(this, void 0);
        _Backbuffer_height.set(this, void 0);
        _Backbuffer_pixelRatio.set(this, void 0);
        _Backbuffer_buffers.set(this, void 0);
        __classPrivateFieldSet(this, _Backbuffer_width, width, "f");
        __classPrivateFieldSet(this, _Backbuffer_height, height, "f");
        __classPrivateFieldSet(this, _Backbuffer_pixelRatio, pixelRatio, "f");
        const pwidth = width * pixelRatio; // use physical size
        const pheight = height * pixelRatio;
        __classPrivateFieldSet(this, _Backbuffer_buffers, [
            new THREE.WebGLRenderTarget(pwidth, pheight),
            new THREE.WebGLRenderTarget(pwidth, pheight),
        ], "f");
    }
    get texture() {
        return __classPrivateFieldGet(this, _Backbuffer_buffers, "f")[0].texture;
    }
    get target() {
        return __classPrivateFieldGet(this, _Backbuffer_buffers, "f")[1];
    }
    /**
     * Resize textuers if necessary.
     * @param width - logical width of the backbuffer
     * @param height - logical height of the backbuffer
     */
    resize(width, height) {
        if (width !== __classPrivateFieldGet(this, _Backbuffer_width, "f") || height !== __classPrivateFieldGet(this, _Backbuffer_height, "f")) {
            __classPrivateFieldSet(this, _Backbuffer_width, width, "f");
            __classPrivateFieldSet(this, _Backbuffer_height, height, "f");
            __classPrivateFieldGet(this, _Backbuffer_buffers, "f")[0].setSize(width * __classPrivateFieldGet(this, _Backbuffer_pixelRatio, "f"), height * __classPrivateFieldGet(this, _Backbuffer_pixelRatio, "f"));
            __classPrivateFieldGet(this, _Backbuffer_buffers, "f")[1].setSize(width * __classPrivateFieldGet(this, _Backbuffer_pixelRatio, "f"), height * __classPrivateFieldGet(this, _Backbuffer_pixelRatio, "f"));
        }
    }
    /**
     * Swap double buffers for the backbuffer.
     * This should always be called right after the rendering.
     */
    swap() {
        __classPrivateFieldSet(this, _Backbuffer_buffers, [__classPrivateFieldGet(this, _Backbuffer_buffers, "f")[1], __classPrivateFieldGet(this, _Backbuffer_buffers, "f")[0]], "f");
    }
    getViewport() {
        return getGLRect(0, 0, __classPrivateFieldGet(this, _Backbuffer_width, "f"), __classPrivateFieldGet(this, _Backbuffer_height, "f"));
    }
}
_Backbuffer_width = new WeakMap(), _Backbuffer_height = new WeakMap(), _Backbuffer_pixelRatio = new WeakMap(), _Backbuffer_buffers = new WeakMap();
