// @vfx-js/core/lib/esm/copy-pass.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _CopyPass_scene, _CopyPass_mesh, _CopyPass_uniforms;
import * as THREE from "three";
import { COPY_FRAGMENT_SHADER, DEFAULT_VERTEX_SHADER } from "./constants.js";
export class CopyPass {
    constructor() {
        _CopyPass_scene.set(this, void 0);
        _CopyPass_mesh.set(this, void 0);
        _CopyPass_uniforms.set(this, void 0);
        __classPrivateFieldSet(this, _CopyPass_scene, new THREE.Scene(), "f");
        __classPrivateFieldSet(this, _CopyPass_uniforms, {
            src: { value: null },
            offset: { value: new THREE.Vector2() },
            resolution: { value: new THREE.Vector2() },
            viewport: { value: new THREE.Vector4() },
        }, "f");
        __classPrivateFieldSet(this, _CopyPass_mesh, new THREE.Mesh(new THREE.PlaneGeometry(2, 2), new THREE.RawShaderMaterial({
            vertexShader: DEFAULT_VERTEX_SHADER,
            fragmentShader: COPY_FRAGMENT_SHADER,
            uniforms: __classPrivateFieldGet(this, _CopyPass_uniforms, "f"),
            glslVersion: "300 es",
            transparent: true,
            premultipliedAlpha: true,
        })), "f");
        __classPrivateFieldGet(this, _CopyPass_scene, "f").add(__classPrivateFieldGet(this, _CopyPass_mesh, "f"));
    }
    get uniforms() {
        return __classPrivateFieldGet(this, _CopyPass_uniforms, "f");
    }
    setUniforms(tex, pixelRatio, xywh) {
        __classPrivateFieldGet(this, _CopyPass_uniforms, "f")["src"].value = tex;
        __classPrivateFieldGet(this, _CopyPass_uniforms, "f")["resolution"].value.x = xywh.w * pixelRatio;
        __classPrivateFieldGet(this, _CopyPass_uniforms, "f")["resolution"].value.y = xywh.h * pixelRatio;
        __classPrivateFieldGet(this, _CopyPass_uniforms, "f")["offset"].value.x = xywh.x * pixelRatio;
        __classPrivateFieldGet(this, _CopyPass_uniforms, "f")["offset"].value.y = xywh.y * pixelRatio;
    }
    get scene() {
        return __classPrivateFieldGet(this, _CopyPass_scene, "f");
    }
}
_CopyPass_scene = new WeakMap(), _CopyPass_mesh = new WeakMap(), _CopyPass_uniforms = new WeakMap();
