// @vfx-js/core/lib/esm/post-effect-pass.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PostEffectPass_scene, _PostEffectPass_mesh, _PostEffectPass_uniforms, _PostEffectPass_uniformGenerators, _PostEffectPass_backbuffer;
import * as THREE from "three";
import { Backbuffer } from "./backbuffer.js";
import { DEFAULT_VERTEX_SHADER } from "./constants.js";
export class PostEffectPass {
    constructor(fragmentShader, uniforms, useBackbuffer) {
        _PostEffectPass_scene.set(this, void 0);
        _PostEffectPass_mesh.set(this, void 0);
        _PostEffectPass_uniforms.set(this, void 0);
        _PostEffectPass_uniformGenerators.set(this, void 0);
        _PostEffectPass_backbuffer.set(this, void 0);
        __classPrivateFieldSet(this, _PostEffectPass_scene, new THREE.Scene(), "f");
        __classPrivateFieldSet(this, _PostEffectPass_uniformGenerators, {}, "f");
        __classPrivateFieldSet(this, _PostEffectPass_uniforms, {
            src: { value: null },
            offset: { value: new THREE.Vector2() },
            resolution: { value: new THREE.Vector2() },
            viewport: { value: new THREE.Vector4() },
            time: { value: 0.0 },
            mouse: { value: new THREE.Vector2() },
        }, "f");
        // Add backbuffer uniform if requested
        if (useBackbuffer) {
            __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f").backbuffer = { value: null };
        }
        // Add custom uniforms if provided
        if (uniforms) {
            for (const [key, value] of Object.entries(uniforms)) {
                if (typeof value === "function") {
                    __classPrivateFieldGet(this, _PostEffectPass_uniformGenerators, "f")[key] = value;
                    __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")[key] = { value: value() };
                }
                else {
                    __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")[key] = { value };
                }
            }
        }
        __classPrivateFieldSet(this, _PostEffectPass_mesh, new THREE.Mesh(new THREE.PlaneGeometry(2, 2), new THREE.RawShaderMaterial({
            vertexShader: DEFAULT_VERTEX_SHADER,
            fragmentShader,
            uniforms: __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f"),
            glslVersion: "300 es",
            transparent: true,
            premultipliedAlpha: true,
        })), "f");
        __classPrivateFieldGet(this, _PostEffectPass_scene, "f").add(__classPrivateFieldGet(this, _PostEffectPass_mesh, "f"));
    }
    get uniforms() {
        return __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f");
    }
    setUniforms(tex, pixelRatio, xywh, time, mouseX, mouseY) {
        __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")["src"].value = tex;
        __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")["resolution"].value.x = xywh.w * pixelRatio;
        __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")["resolution"].value.y = xywh.h * pixelRatio;
        __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")["offset"].value.x = xywh.x * pixelRatio;
        __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")["offset"].value.y = xywh.y * pixelRatio;
        __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")["time"].value = time;
        __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")["mouse"].value.x = mouseX * pixelRatio;
        __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")["mouse"].value.y = mouseY * pixelRatio;
    }
    updateCustomUniforms(uniformGenerators) {
        // Update uniforms from constructor generators
        for (const [key, generator] of Object.entries(__classPrivateFieldGet(this, _PostEffectPass_uniformGenerators, "f"))) {
            if (__classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")[key]) {
                __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")[key].value = generator();
            }
        }
        // Update uniforms from external generators (for compatibility)
        if (uniformGenerators) {
            for (const [key, generator] of Object.entries(uniformGenerators)) {
                if (__classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")[key]) {
                    __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f")[key].value = generator();
                }
            }
        }
    }
    initializeBackbuffer(width, height, pixelRatio) {
        if (__classPrivateFieldGet(this, _PostEffectPass_uniforms, "f").backbuffer && !__classPrivateFieldGet(this, _PostEffectPass_backbuffer, "f")) {
            __classPrivateFieldSet(this, _PostEffectPass_backbuffer, new Backbuffer(width, height, pixelRatio), "f");
            __classPrivateFieldGet(this, _PostEffectPass_uniforms, "f").backbuffer.value = __classPrivateFieldGet(this, _PostEffectPass_backbuffer, "f").texture;
        }
    }
    resizeBackbuffer(width, height) {
        if (__classPrivateFieldGet(this, _PostEffectPass_backbuffer, "f")) {
            __classPrivateFieldGet(this, _PostEffectPass_backbuffer, "f").resize(width, height);
        }
    }
    get backbuffer() {
        return __classPrivateFieldGet(this, _PostEffectPass_backbuffer, "f");
    }
    get scene() {
        return __classPrivateFieldGet(this, _PostEffectPass_scene, "f");
    }
}
_PostEffectPass_scene = new WeakMap(), _PostEffectPass_mesh = new WeakMap(), _PostEffectPass_uniforms = new WeakMap(), _PostEffectPass_uniformGenerators = new WeakMap(), _PostEffectPass_backbuffer = new WeakMap();
