// @vfx-js/core/lib/cjs/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVFXOpts = getVFXOpts;
/**
 * Parse VFXOpts and fill in the default values.
 * @internal
 */
function getVFXOpts(opts) {
    var _a, _b, _c, _d, _e;
    const defaultPixelRatio = typeof window !== "undefined" ? window.devicePixelRatio : 1;
    let scrollPadding;
    if (opts.scrollPadding === undefined) {
        scrollPadding = [0.1, 0.1];
    }
    else if (opts.scrollPadding === false) {
        scrollPadding = [0, 0];
    }
    else if (Array.isArray(opts.scrollPadding)) {
        scrollPadding = [
            (_a = opts.scrollPadding[0]) !== null && _a !== void 0 ? _a : 0.1,
            (_b = opts.scrollPadding[1]) !== null && _b !== void 0 ? _b : 0.1,
        ];
    }
    else {
        scrollPadding = [opts.scrollPadding, opts.scrollPadding];
    }
    return {
        pixelRatio: (_c = opts.pixelRatio) !== null && _c !== void 0 ? _c : defaultPixelRatio,
        zIndex: (_d = opts.zIndex) !== null && _d !== void 0 ? _d : undefined,
        autoplay: (_e = opts.autoplay) !== null && _e !== void 0 ? _e : true,
        fixedCanvas: opts.scrollPadding === false,
        scrollPadding,
        postEffect: opts.postEffect,
    };
}
